rfcs/text/0011_persistable_state_service.md

```diff
@@ -0,0 +1,340 @@
+      register: (id: string, definition: Partial<PersistableStateDefinition>) => void
+    }
+  }
+  start(core, plugins) {
+    return {
+      // get will always return. If requested persistable state is not defined it will return default set of functions.
+      get: (id: string) => PersistableStateDefinition,
+      // takes the state, extracts the references and returns them + version string
+      beforeSave: (id: string, state: PersistableState) => [PersistableState, SavedObjectReference[], string],
+      // takes the state, references and version and returns latest (migrated) state with references injected
+      afterLoad: (id: string, state: PersistableState, references: SavedObjectReference[], version: string) => PersistableState,
+    }
+  }
+}
+```
+
+Every plugin that exposes its state should register its `PersistableStateDefinition`. In this definition he needs to expose `migrate` function which can take in any state version and migrate it to the latest state as well as `inject` and `extract` functions, which should be able to inject and extract saved object references from the latest state version.
+
+This same persitableStateDefinition should also be exposed on the plugin contract. On top of that plugin also needs to export type definition for the latest state version. Name of type definition should not include a version number but always be the same. (see the basic example)
+
+Plugins consuming state not owned by them should always use persistableStateDefinion of the registrator plugin. When loading state they need to:
+
+`const state = persistabelStateDefinition.inject(persistableStateDefinition.migrate(loadedState, version), references)`
+
+and when saving state they always need to:
+
+`const { state, references } = persistableStateDefinition.extract(runtimeState)` and store references separatly.
+
+This way consuming plugin can always be sure that its working with the latest state version.
+
+# Server or Client ?
+
+Persistable state service will be available on both, server and client, under same interface. We suggest to register your persistable state definition on both client and server even if you are using just one of those as you cannot know where the consuming side is running.
+
+
+# Consuming and edge cases
+
+When plugin A wants to store some state that it does not own it should check with persistableStateRegistry for a registered persistable state definition and execute `beforeSave` function.
+
+```ts
+  const stateForSaving = persistableStateRegistry.beforeSave(id, myState);
+
+``` 
+
+WARNING: If state id is known at compile time we should rather import the correct utilities directly from the registrator plugin. As using the registry can hide dependencies between plugins when possible plugins should directly depend on plugins providing the state they want to store.
+
+```ts
+  import { extract } from 'src/plugins/pluginOwningState';
+  const stateForSaving = extract(myState);
+```
```

Comment:
Why the static import here, can't this be achieved using setup contract instead?

Also: 

> If state id is known at compile time we should rather import the correct utilities directly from the registrator plugin

As it's supposed to be working using the global registry anyway, what about 'prefer one way to do things' ?

Reply:
using global registry when you know the id of the state you are working with at compile time is not desired as introduces indirect dependencies (you define dependency of persistable state service, but in reality you depend on pluginOwningState, so you should rather work with plugin owning state directly.

wheter the methods are exposed on the contract or staticly is imo implementation detail of plugin exposing them, i think static exports are prefered when possible but i don't think that should be defined here.